runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/GreedilyFusedExecutableStageTest.java

```diff
@@ -0,0 +1,747 @@
+                .build());
+    Set<PTransformNode> differentEnvironments = p.getPerElementConsumers(readOutputNode);
+
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("go");
+    thrown.expectMessage("py");
+    thrown.expectMessage("same");
+    GreedilyFusedExecutableStage.forGrpcPortRead(p, readOutputNode, differentEnvironments);
+  }
+
+  @Test
+  public void noEnvironmentThrows() {
+    QueryablePipeline p =
+        QueryablePipeline.fromComponents(
+            Components.newBuilder()
+                .putTransforms(
+                    "read", PTransform.newBuilder().putOutputs("output", "read.out").build())
+                .putPcollections("read.out", readDotOut)
+                .putTransforms(
+                    "runnerTransform",
+                    PTransform.newBuilder()
+                        .putInputs("input", "read.out")
+                        .setSpec(
+                            FunctionSpec.newBuilder()
+                                .setUrn(PTransformTranslation.GROUP_BY_KEY_TRANSFORM_URN))
+                        .putOutputs("output", "gbk.out")
+                        .build())
+                .putPcollections(
+                    "gbk.out", PCollection.newBuilder().setUniqueName("gbk.out").build())
+                .build());
+    Set<PTransformNode> differentEnvironments = p.getPerElementConsumers(readOutputNode);
+
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("Environment must be populated");
+    GreedilyFusedExecutableStage.forGrpcPortRead(p, readOutputNode, differentEnvironments);
+  }
+
+  @Test
+  public void nonReadRootThrows() {
+    QueryablePipeline p =
+        QueryablePipeline.fromComponents(
+            Components.newBuilder()
+                .putTransforms(
+                    "rootParDo",
+                    PTransform.newBuilder()
+                        .setSpec(
+                            FunctionSpec.newBuilder()
+                                .setUrn(PTransformTranslation.PAR_DO_TRANSFORM_URN)
+                                .setPayload(
+                                    ParDoPayload.newBuilder()
+                                        .setDoFn(
+                                            SdkFunctionSpec.newBuilder().setEnvironmentId("env"))
+                                        .build()
+                                        .toByteString()))
+                        .putOutputs("output", "read.out")
+                        .build())
+                .putPcollections("read.out", readDotOut)
+                .putEnvironments("env", Environment.newBuilder().setUrl("env").build())
+                .build());
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage(PTransformTranslation.READ_TRANSFORM_URN);
+    GreedilyFusedExecutableStage.forRootTransform(p, getOnlyElement(p.getRootTransforms()));
+  }
+
+  @Test
+  public void fusesCompatibleEnvironmentsWithRead() {
```

Comment:
Its useful for all the tests to provide a little ascii diagram of the transforms/pcollections like:
`read --> read.out --> parDo --> parDo.out`

Reply:
Done for all, with little explanations if appropriate.